### ISSUE TITLE
Search: edX returns 401 error on enrollments API

### ISSUE THREAD

**Issue by pdpinch** [mitodl/micromasters#2132]
(this is a search issue because search results won't be accurate without accurate enrollment data)

This is happening in production, but only a few times. I don't think there's any user impact, because the server should simply re-authenticate with the refresh token, but it would be good to confirm and  quiet the error. 

https://sentry.io/mit-office-of-digital-learning/micromasters/issues/193079253/

```
HTTPError: 401 Client Error: UNAUTHORIZED for url: https://courses.edx.org/api/enrollment/v1/enrollment
  File "dashboard/tasks.py", line 137, in batch_update_user_data_subtasks
    CachedEdxDataApi.update_cache_if_expired(user, edx_client, cache_type)
  File "dashboard/api_edx_cache.py", line 282, in update_cache_if_expired
    update_func(user, edx_client)
  File "dashboard/api_edx_cache.py", line 165, in update_cached_enrollments
    enrollments = edx_client.enrollments.get_student_enrollments()
  File "edx_api/enrollments/__init__.py", line 36, in get_student_enrollments
    resp.raise_for_status()
  File "requests/models.py", line 862, in raise_for_status
    raise HTTPError(http_error_msg, response=self)

Unable to refresh cache enrollment for student NimishSharma
```

**Comment by giocalitri** [comment]
if the error happens during the token refresh, the solution is to re-login the user (the following is a test with Jared's user
```
# before the login via web page:
>>> from django.contrib.auth.models import User
>>> from backends.edxorg import EdxOrgOAuth2
>>> user = User.objects.get(username='jtausig')
>>> user_social = user.social_auth.get(provider=EdxOrgOAuth2.name)
>>> user_social
<UserSocialAuth: jtausig>
>>> from social.apps.django_app.utils import load_strategy
>>> strategy = load_strategy()
>>> user_social.refresh_token(strategy)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/app/.heroku/python/lib/python3.5/site-packages/social/storage/base.py", line 54, in refresh_token
    response = backend.refresh_token(token, *args, **kwargs)
  File "/app/backends/edxorg.py", line 108, in refresh_token
    response = super(EdxOrgOAuth2, self).refresh_token(token, *args, **kwargs)
  File "/app/.heroku/python/lib/python3.5/site-packages/social/backends/oauth.py", line 430, in refresh_token
    request = self.request(url, **request_args)
  File "/app/.heroku/python/lib/python3.5/site-packages/social/backends/base.py", line 225, in request
    response.raise_for_status()
  File "/app/.heroku/python/lib/python3.5/site-packages/requests/models.py", line 862, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: UNAUTHORIZED for url: https://courses.edx.org/oauth2/access_token/


# this is after a login via the MicroMasters web page
>>> user_social.refresh_from_db()
>>> user_social.refresh_token(strategy)
```

Looking if the user in the issue has the same problem

**Comment by giocalitri** [comment]
same error for the user in the sentry error 
```
>>> user = User.objects.get(username='Cv_vignesh')
>>> user_social = user.social_auth.get(provider=EdxOrgOAuth2.name)
>>> user_social
<UserSocialAuth: Cv_vignesh>
>>> user_social.refresh_token(strategy)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/app/.heroku/python/lib/python3.5/site-packages/social/storage/base.py", line 54, in refresh_token
    response = backend.refresh_token(token, *args, **kwargs)
  File "/app/backends/edxorg.py", line 108, in refresh_token
    response = super(EdxOrgOAuth2, self).refresh_token(token, *args, **kwargs)
  File "/app/.heroku/python/lib/python3.5/site-packages/social/backends/oauth.py", line 430, in refresh_token
    request = self.request(url, **request_args)
  File "/app/.heroku/python/lib/python3.5/site-packages/social/backends/base.py", line 225, in request
    response.raise_for_status()
  File "/app/.heroku/python/lib/python3.5/site-packages/requests/models.py", line 862, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: UNAUTHORIZED for url: https://courses.edx.org/oauth2/access_token/
```

**Comment by giocalitri** [comment]
for the user in the ticket, it works now, so my assumption is that s/he was able to re-login and fix the problem.

```
>>> user = User.objects.get(username='NimishSharma')
>>> user_social = user.social_auth.get(provider=EdxOrgOAuth2.name)
>>> user_social
<UserSocialAuth: NimishSharma>
>>> user_social.refresh_token(strategy)
>>> from edx_api.client import EdxApi
>>> from django.conf import settings
>>> edx_client = EdxApi(user_social.extra_data, settings.EDXORG_BASE_URL)
>>> enrollment = 'enrollment'
>>> from dashboard.api_edx_cache import CachedEdxDataApi
>>> CachedEdxDataApi.update_cache_if_expired(user, edx_client, enrollment)
>>> from dashboard.models import UserCacheRefreshTime
>>> UserCacheRefreshTime.objects.get(user=user)
<UserCacheRefreshTime: user "NimishSharma">
>>> update_timestamps = UserCacheRefreshTime.objects.get(user=user)
>>> update_timestamps.enrollment
datetime.datetime(2017, 1, 3, 21, 30, 24, 59742, tzinfo=<UTC>)
>>> update_timestamps.certificate
datetime.datetime(2017, 1, 3, 18, 31, 15, 190214, tzinfo=<UTC>)
>>> update_timestamps.current_grade
datetime.datetime(2017, 1, 3, 18, 31, 23, 615024, tzinfo=<UTC>)
```

**Comment by giocalitri** [comment]
Now the problem is to figure out why the user got at the point where s/he was not required to refresh the token (or s/he refreshed successfully, even if I doubt this happened) but s/he was unauthorized on edX.

**Comment by pdpinch** [comment]
@giocalitri I feel like I'm seeing this happen more often, and in the case of one user it led to an error on the dashboard page. Can we make sure that the user is asked to log in again when this happens?

user 1073 raised this error in Sentry:

https://sentry.io/mit-office-of-digital-learning/micromasters/issues/193330670/events/4896741035/

and saw this error on their dashboard:`500 Sorry, we were unable to load the data necessary to process your request. Please reload the page.
`

**Comment by pdpinch** [comment]
https://sentry.io/mit-office-of-digital-learning/micromasters/issues/182642956/

This is actually a different endpoint (not enrollments) but we think it's the same root cause. 

To be clear, I'm most interested in solving this for live users, not the cache updates done by the workers.

**Comment by pdpinch** [comment]
https://sentry.io/mit-office-of-digital-learning/micromasters/issues/193330670/

**Comment by pdpinch** [comment]
@giocalitri what more do we need to do here now that #2709 has been merged and released?

**Comment by pdpinch** [comment]
https://sentry.io/mit-office-of-digital-learning/micromasters/issues/228392933/

**Comment by pdpinch** [comment]
@giocalitri is there a way we can estimate how widespread this is? In other words, for how many users are we in doubt about their grades and enrollments?

**Comment by giocalitri** [comment]
the only way to be sure is to run a script and try to refresh the token for each student

an approximation might be the number of exceptions you see in sentry at a given time (that should correspond to a run of the batch refresh)

**Comment by pdpinch** [comment]
Giovanni realized that we can estimate this just by looking at the refresh dates for cached data. It's supposed to refresh every 6 hours. Anything that isn't refreshed is a user who's token isn't working. There are 400-500 of these now I think.

**Comment by pdpinch** [comment]
Did we learn anything more from edX about this users?

**Comment by pdpinch** [comment]
@giocalitri Did we learn anything more from edX about this users?

**Comment by giocalitri** [comment]
No, Joel never updated me: I will contact him.

**Comment by pdpinch** [comment]
Ok. I'd like to get your thoughts on next steps, after you've heard from him. 

Some things we talked about:

- e-mail all the users with expired tokens & ask them to log in
- failover to a system account that can update grades (but not enrollments or certificates)
- failover to institutional research's data in bigquery (blocked due to authorization issues, but could be addressed)

**Comment by giocalitri** [comment]
don't we want to try to modify openedx to allow the system account to access also the enrollments and certificates?

**Comment by pdpinch** [comment]
further testing has suggested that this happens when users change their passwords on edX, which invalidates their OAuth tokens. Unless and until they return to the portal, we won't be able to update their enrollments or grades.

**Comment by pdpinch** [comment]
This can be closed with #3700 merges.
